### PR TITLE
docs: clarify Vault CA provider permissions needed

### PR DESCRIPTION
**Pull request by jkirschner-hashicorp**
After this PR is merged, the associated tutorials that include Vault policies for use with the Vault as a Connect CA provider must be updated as well.

[Preview link to the Vault policies changes](https://consul-n4pht8hnd-hashicorp.vercel.app/consul/docs/connect/ca/vault#vault-acl-policies)

### CODE REVIEW

File: website/content/docs/connect/ca/vault.mdx
Comment (jkirschner-hashicorp) [inline website/content/docs/connect/ca/vault.mdx:231]: Question for reviewer: do we think we should move the auth token renewal stuff to the bottom of the list of permissions? Maybe the most important stuff is the CA stuff itself.

(I think I agree with this. Will probably move to bottom unless told otherwise.)

File: website/content/docs/connect/ca/vault.mdx
Comment (jkirschner-hashicorp) [inline website/content/docs/connect/ca/vault.mdx:219]: Note to self: I'm still missing the guidance on what permissions are needed to change the provider away from the Vault CA provider.

That should probably be mentioned in the CLI/HTTP API docs as well... I'll address that in a future commit.

File: website/content/docs/connect/ca/vault.mdx
Comment (jkirschner-hashicorp) [inline website/content/docs/connect/ca/vault.mdx:223]: As of [this PR](https://github.com/hashicorp/consul/pull/12655/files#diff-7bf30611a760296e2e6ffcd7fe955b0eb8317fe0a43c470b8b52d311af4a2515L359), I don't think we actually need read capabilities on `sys/mounts` anymore... There's no other call to `Sys().ListMounts()` as far as I can tell. (We should actually test this first.)

Can we remove that from the docs? Or at least mark the version upon which it is no longer necessary?

Comment (jkirschner-hashicorp) [reply website/content/docs/connect/ca/vault.mdx:223]: It looks like this change was made in Consul 1.12.0: https://github.com/hashicorp/consul/blob/main/CHANGELOG.md#1120-april-20-2022

File: website/content/docs/connect/ca/vault.mdx
Comment (jkirschner-hashicorp) [inline website/content/docs/connect/ca/vault.mdx:283]: As of [this PR](https://github.com/hashicorp/consul/pull/12655/files#diff-7bf30611a760296e2e6ffcd7fe955b0eb8317fe0a43c470b8b52d311af4a2515L359), I don't think we actually need read capabilities on sys/mounts anymore... There's no other call to Sys().ListMounts() as far as I can tell. (We should actually test this first.)


It looks like this change was made in Consul 1.12.0: https://github.com/hashicorp/consul/blob/main/CHANGELOG.md#1120-april-20-2022

File: website/content/docs/connect/ca/vault.mdx
Comment (jkirschner-hashicorp) [inline website/content/docs/connect/ca/vault.mdx:264]: @rboyer : This is my attempt to strike a balance between documenting this necessary permission while optimizing for security.

I also reference this from the "modify CA" CLI and HTTP API docs.

File: website/content/commands/connect/ca.mdx
Comment (jkirschner-hashicorp) [inline website/content/commands/connect/ca.mdx:106]:
```suggestion
   [`<insert-RootPKIPath>/root/sign-self-issued` Vault endpoint](/vault/api-docs/secret/pki#sign-self-issued)
```

**Review by rboyer — APPROVED** [review]

File: website/content/docs/connect/ca/vault.mdx
Comment (kisunji) [inline website/content/docs/connect/ca/vault.mdx:226]: Since the `<>` already suggests "replace me" can we use the documented config field names:
```suggestion
path "/sys/mounts/<root_pki_path>" {
```

File: website/content/docs/connect/ca/vault.mdx
Comment (kisunji) [reply website/content/docs/connect/ca/vault.mdx:283]: Can this be handled with versioned docs? I found the commented versioning guidelines for `/tune` confusing.

**Review by kisunji — APPROVED** [review]

File: website/content/docs/connect/ca/vault.mdx
Comment (kisunji) [reply website/content/docs/connect/ca/vault.mdx:283]: I support removing version references from config code blocks and calling out 1.12+ required 👍

File: website/content/docs/connect/ca/vault.mdx
Comment (trujillo-adam) [inline website/content/docs/connect/ca/vault.mdx:214]:
```suggestion
The `RootPKIPath` and `IntermediatePKIPath` paths have placeholder values. Replace them with real path values in your CA provider configuration.
```
In docs, fewer words outweighs politeness.

File: website/content/docs/connect/ca/vault.mdx
Comment (trujillo-adam) [inline website/content/docs/connect/ca/vault.mdx:219]: Curious about these in-file explanations. Do they mirror a template file included in some package?

**Review by trujillo-adam — CHANGES_REQUESTED** [review]
Can we discuss the docs-within-the-docs approach here? Are we mirroring a template file shipped in some package or repository? If not, I don't think we should suddenly start providing documentation inside code snippets. Tagging @boruszak @im2nguyen for awarenss

File: website/content/docs/connect/ca/vault.mdx
Comment (jkirschner-hashicorp) [reply website/content/docs/connect/ca/vault.mdx:219]: I didn't check what established practices we have for including comments in HCL file examples. JSON doesn't support comments, so it's not even possible in that context.

I included these comments so that, if someone copies and pastes this policy for their own usage, someone looking at it now or later can understand the purpose when it might not otherwise be obvious. In my experience, I've seen configuration file templates included with software include inline docs comments like this, including commented out things with an explanation of when it should be uncommented (like I've done here for the root rotation permission).

I'm open to removing the comments. I personally found the policy a lot more understandable with them than without them. It also helps with playing "spot the difference" between the Consul vs Vault managed mounts approaches.

Before proceeding, I'm curious what @boruszak and @im2nguyen think.

Comment (trujillo-adam) [reply website/content/docs/connect/ca/vault.mdx:219]: Here is more context about how I'm looking at the comment-based docs.

First, I do see the value. I have documented systems where the product ships with a commented up configuration file that the user duplicates and modifies to their implementation. It can be a useful way to present the information. 

For me, there are a few issues. As you pointed out, you can only provide the comments in HCL (and YAML), so we can't provide the examples for JSON configurations. We lose the rich text features, such as hyperlinks and text formatting, which might be useful in the explanations. Third, it's a deviation from our convention. Are users now expecting documentation within other example configuration files? 

I think the big thing, though, is that the information in this section is presented as a hybrid of conceptual and reference when we should present it as usage. There appears to be logical splits in the policy file where we could create "Define Vault managed PKI mount policies" section, for example. The section would include an explanation with links to the configuration reference items and an example. Or maybe it's a step-by-step process describing how to configure blocks of rules. When the information is presented as conceptual or reference, I think it reads as less actionable, so readers don't know if they should or have to do something.

I'll let Jeff and Tu chime in, though.

Comment (jkirschner-hashicorp) [reply website/content/docs/connect/ca/vault.mdx:219]: If we went with the framing of "leave in comments where something may require thought/action", the only comment we'd want to leave in as the one about enabling root rotation, since best practices would suggest granting that permission for the minimum amount of time necessary when rotation is needed.

File: website/content/docs/connect/ca/vault.mdx
Comment (jkirschner-hashicorp) [reply website/content/docs/connect/ca/vault.mdx:214]: Modified - check the latest wording

File: website/content/docs/connect/ca/vault.mdx
Comment (jkirschner-hashicorp) [inline website/content/docs/connect/ca/vault.mdx:187]: This headings are often linked to. We need to create a link alias to avoid breaking existing links.

File: website/content/docs/connect/ca/vault.mdx
Comment (jkirschner-hashicorp) [inline website/content/docs/connect/ca/vault.mdx:195]: Note to self: go back through all of these and make sure root vs intermediate is used in all the right places.

File: website/content/docs/connect/ca/vault.mdx
Comment (jkirschner-hashicorp) [inline website/content/docs/connect/ca/vault.mdx:260]: This headings are often linked to. We need to create a link alias to avoid breaking existing links.

File: website/content/docs/connect/ca/vault.mdx
Comment (jkirschner-hashicorp) [inline website/content/docs/connect/ca/vault.mdx:179]: The original statement wasn't clear enough. The suggestion replacement isn't logically equivalent, though. I'll need to update.

File: website/content/docs/connect/ca/vault.mdx
Comment (jkirschner-hashicorp) [reply website/content/docs/connect/ca/vault.mdx:221]: After checking, I fixed one mistake where I mixed up root and intermediate.

**Comment by jkirschner-hashicorp** [comment]
@trujillo-adam : ready for re-review

File: website/content/docs/connect/ca/vault.mdx
Comment (trujillo-adam) [inline website/content/docs/connect/ca/vault.mdx:213]:
```suggestion
Next, define a Vault ACL policy that allows Consul to use the pre-existing PKI mounts in Vault.
The policy grants Consul limited use of the root PKI engine
and full control of the intermediate PKI engine to issue service mesh certificates.
```
Aren't we repeating the same information in the steps below? I think we can delete this paragraph and the one above it. The next paragraph is a good introduction to the procedure for building a suitable policy.

File: website/content/docs/connect/ca/vault.mdx
Comment (trujillo-adam) [inline website/content/docs/connect/ca/vault.mdx:284]:
```suggestion
1. You should only include the following rules in your Vault policy file
   while changing the CA provider from Vault to a different CA provider, such as Consul's built-in provider,
   or changing the `RootPKIPath` used by the Vault CA provider.
   The CA provider modifications trigger a root CA change that requires a
   highly-privileged root cross-sign operation.
   Remove the elevated privilege immediately after the CA provider modification is finished.
```
I get why you wanted to label the step, but it adds too many words without as much benefit. As far as the recommendation to remove the elevated privileges, should that be a final step at the end of this procedure? It would make this step shorter and provide a dedicated action for readers following the process.

File: website/content/docs/connect/ca/vault.mdx
Comment (trujillo-adam) [inline website/content/docs/connect/ca/vault.mdx:297]: Same comment as the previous procedure - we already describe what each rule in the policy does in the steps, so I think we should remove the first two paragraphs in this introduction.

File: website/content/docs/connect/ca/vault.mdx
Comment (trujillo-adam) [inline website/content/docs/connect/ca/vault.mdx:367]:
```suggestion
1. You should only include the following rules in your Vault policy file
   while you are changing the CA provider from Vault to a different CA provider, such as Consul's built-in provider,
   or changing the `RootPKIPath` used by the Vault CA provider.
   The CA provider modifications trigger a root CA change that requires a
  highly-privileged root cross-sign operation.
   Remove the elevated privilege immediately after the CA provider modification is finished.
```
Same comment as above

**Review by trujillo-adam — COMMENTED** [review]
I added a few more suggestions, mostly to add more precise language where necessary and to streamline areas where I think we're over-explaining.

File: website/content/api-docs/connect/ca.mdx
Comment (jkirschner-hashicorp) [reply website/content/api-docs/connect/ca.mdx:199]: @im2nguyen : On a separate PR, I received feedback that we want to standardize on generic callouts like "Note / Caution / Warning" for the different callout types. It seems like we may not be internally aligned on which direction we want to go between specific callouts vs general callouts.

File: website/content/docs/connect/ca/vault.mdx
Comment (jkirschner-hashicorp) [reply website/content/docs/connect/ca/vault.mdx:22]: They are both Consul options/configurations. I also find the correction to be a bit confusing. I'd prefer to keep the additional wording.

Regarding the call-out text, same comment as above: I'll go with whatever you and @im2nguyen agree upon on generic vs specific.

File: website/content/docs/connect/ca/vault.mdx
Comment (jkirschner-hashicorp) [reply website/content/docs/connect/ca/vault.mdx:28]: I hesitated to write it this way to begin with because their are multiple types of files that can contain configuration ... "configuration entries" are different from "agent configuration".

That said, perhaps it's better overall to make this suggested change, even though there may be a possibility of misinterpretation.

File: website/content/docs/connect/ca/vault.mdx
Comment (jkirschner-hashicorp) [reply website/content/docs/connect/ca/vault.mdx:72]: At some point, we should just make a pass over all the pages to change the capitalization.

For any _new_ headers I create, I'm using the style guide. But for content I'm not editing (or only shuffling on the page), I haven't been touching the existing header to make them compliant (because I'm focused on a particular change).

This probably creates a lot of back and forth for both reviewers and authors. If we update all the content at once, it should be easier for everyone going forward.

File: website/content/docs/connect/ca/vault.mdx
Comment (jkirschner-hashicorp) [reply website/content/docs/connect/ca/vault.mdx:213]: Yeah. I think I wrote this before I had the summary paragraph above that compares the two options.

File: website/content/docs/connect/ca/vault.mdx
Comment (jkirschner-hashicorp) [reply website/content/docs/connect/ca/vault.mdx:284]: The policy snippets in steps 1-3 must **always** be in place for Consul to work.

This policy snippet is ideally only in place for a very brief period of time while performing a specific operation.

These four "steps" aren't really a _procedure_ to follow. I don't think it makes sense to add a "step 5" for removing the elevated privileges.

Maybe the right solution is to move this temporarily privilege into its own section? It's shared between Vault and Consul managed paths, so I don't need to duplicate it in both sections... I'll try that out so we can take a look.

File: website/content/api-docs/connect/ca.mdx
Comment (jkirschner-hashicorp) [reply website/content/api-docs/connect/ca.mdx:199]: Use "caution"

File: website/content/docs/connect/ca/vault.mdx
Comment (jkirschner-hashicorp) [reply website/content/docs/connect/ca/vault.mdx:22]: "Compatibility warning"

File: website/content/api-docs/connect/ca.mdx
Comment (jkirschner-hashicorp) [inline website/content/api-docs/connect/ca.mdx:172]: Note to self: revisit this before submitting for final review

File: website/content/commands/connect/ca.mdx
Comment (jkirschner-hashicorp) [inline website/content/commands/connect/ca.mdx:100]: Note to self: revisit before final review

File: website/content/docs/connect/ca/vault.mdx
Comment (jkirschner-hashicorp) [reply website/content/docs/connect/ca/vault.mdx:22]: This paragraph will also be touched again in about ~1 week, as we're actively working to remediate this incompatibility.

File: website/content/docs/connect/ca/vault.mdx
Comment (jkirschner-hashicorp) [inline website/content/docs/connect/ca/vault.mdx:359]: @rboyer : In practice... how would a Vault user actually accomplish this?

If I understand correctly, a Vault token's policies are immutable. That would mean you'd need to:
- Create a new Vault token with the elevated policy
- Apply the new Vault token to the Consul Connect CA configuration
- Perform the operation
- Reapply the previous Vault token with non-elevated permissions to the Consul Connect CA configuration

Alternatively, you could use a Vault auth method, and modify the policy associated with the auth method... But if I understand correctly, that wouldn't actually take effect unless Consul re-logs in using the auth method (after the modification) to get a new token (which contains the modified rules). And I don't know how the user would force Consul to do that... So maybe this is only workable when using a Vault token directly rather than an auth method, even though an auth method is generally preferable for enhanced security?

**Review by trujillo-adam — APPROVED** [review]
There are still some language and style guide issues, but I want to unblock you from getting the correct information published. I can open a separate PR to address the writing issues.

**Pull request merged**
